File: CPP/Algorithm/NOD_&_SOD.cpp
// Complexity: O(n * log(n))
// Application: Used to determine the Number of Divisors and Sum of Divisors from number 1 to n.

long long nd[N];
void pre() {
    for (int i = 1; i < N; ++i)
        for (int j = i; j < N; j += i) ++nd[j];
}

long long sd[N];
void pre() {
    for (int i = 1; i < N; ++i)
        for (int j = i; j < N; j += i) sd[j] += i;
}

File: Python/Algorithm/NOD_&_SOD.py
# Complexity: O(n * log(n))
# Application: Used to determine the Number of Divisors and Sum of Divisors from number 1 to n.

def nod(n):
    nd = [0] * (n + 1)
    for i in range(1, n + 1):
        for j in range(i, n + 1, i): nd[j] += 1
    return nd

def sod(n):
    sd = [0] * (n + 1)
    for i in range(1, n + 1):
        for j in range(i, n + 1, i): sd[j] += i
    return sd